src/app-layer-htp-file.c

```diff
@@ -332,17 +332,53 @@ int HTPFileOpenWithRange(HtpState *s, const uint8_t *filename, uint16_t filename
         return HTPFileOpen(
                 s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
     }
-    s->file_range = ContainerUrlRangeOpenFile(file_range_container, crparsed.start, crparsed.end,
+    DEBUG_VALIDATE_BUG_ON(s->file_range);
+    s->file_range = HttpRangeOpenFile(file_range_container, crparsed.start, crparsed.end,
             crparsed.size, &s->cfg->response.sbcfg, filename, filename_len, flags, data, data_len);
+    SCLogDebug("s->file_range == %p", s->file_range);
     if (s->file_range == NULL) {
+        SCLogDebug("s->file_range == NULL");
+        THashDecrUsecnt(file_range_container->hdata);
+        DEBUG_VALIDATE_BUG_ON(
+                SC_ATOMIC_GET(file_range_container->hdata->use_cnt) > (uint32_t)INT_MAX);
+        THashDataUnlock(file_range_container->hdata);
+
         // probably reached memcap
         return HTPFileOpen(
                 s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
+        /* in some cases we don't take a reference, so decr use cnt */
+    } else if (s->file_range->container == NULL) {
+        THashDecrUsecnt(file_range_container->hdata);
+    } else {
+        SCLogDebug("container %p use_cnt %u", s->file_range,
+                SC_ATOMIC_GET(s->file_range->container->hdata->use_cnt));
+        DEBUG_VALIDATE_BUG_ON(
+                SC_ATOMIC_GET(s->file_range->container->hdata->use_cnt) > (uint32_t)INT_MAX);
     }
 
+    /* we're done, so unlock. But since we have a reference in s->file_range keep use_cnt. */
+    THashDataUnlock(file_range_container->hdata);
     SCReturnInt(0);
 }
 
+static void HTPFileStoreChunkHandleRange(
+        HttpRangeContainerBlock *c, const uint8_t *data, uint32_t data_len)
+{
+    if (c->container) {
+        THashDataLock(c->container->hdata);
```

Comment:
Why would we need this lock ?

The point of my architecture was to have the concurrency only in open and close, so that append is simply protected by the lock of its flow.
To say it another way, we should be able to append data to 2 different buffers at the same time.
Am I missing something ?

Reply:
The idea is that any access to the shared data, so `HttpRangeContainerFile` and everything inside it, needs to happen under lock as we can have multiple flows trying to work with it at the same time.

When we create a new range the hash table returns a locked `HttpRangeContainerFile` to us. For any updates following that we need to lock it on append and close directly.

Reply:
I will read this again in the morning right after ☕

Reply:
Maybe we need a live talk about it.

> so HttpRangeContainerFile and everything inside it, needs to happen under lock as we can have multiple flows trying to work with it at the same time.

I totally agree.

But `HTPFileStoreChunkHandleRange ` is about `HttpRangeContainerBlock`, not `HttpRangeContainerFile` which I think does not need a lock, as it is accessed by only one flow.
`HttpRangeContainerBlock` just checks its `container` filled, as you are doing here without a lock, and only one of them is allowed to `FileAppendData(c->container->files, data, len)` cf `appending` set in `HttpRangeOpenFileAux
It is not at risk by the timeout since timeout will occur only if the use count is zero (no more flows have a reference to the`HttpRangeContainerFile`)